// nylas/lib/models/event-conferencing.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventConferencing = void 0;
var restful_model_1 = __importDefault(require("./restful-model"));
var attributes_1 = __importDefault(require("./attributes"));
var EventConferencingDetails = /** @class */ (function (_super) {
    __extends(EventConferencingDetails, _super);
    function EventConferencingDetails() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    EventConferencingDetails.prototype.toJSON = function () {
        return {
            meeting_code: this.meetingCode,
            phone: this.phone,
            password: this.password,
            pin: this.pin,
            url: this.url,
        };
    };
    return EventConferencingDetails;
}(restful_model_1.default));
EventConferencingDetails.collectionName = 'event_conferencing_details';
EventConferencingDetails.attributes = {
    meetingCode: attributes_1.default.String({
        modelKey: 'meetingCode',
        jsonKey: 'meeting_code',
    }),
    phone: attributes_1.default.StringList({
        modelKey: 'phone',
    }),
    password: attributes_1.default.String({
        modelKey: 'password',
    }),
    pin: attributes_1.default.String({
        modelKey: 'pin',
    }),
    url: attributes_1.default.String({
        modelKey: 'url',
    }),
};
var EventConferencing = /** @class */ (function (_super) {
    __extends(EventConferencing, _super);
    function EventConferencing() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    EventConferencing.prototype.toJSON = function () {
        return {
            details: this.details,
            provider: this.provider,
            autocreate: this.autocreate,
        };
    };
    return EventConferencing;
}(restful_model_1.default));
exports.EventConferencing = EventConferencing;
EventConferencing.collectionName = 'event_conferencing';
EventConferencing.attributes = {
    details: attributes_1.default.Object({
        modelKey: 'details',
        itemClass: EventConferencingDetails,
    }),
    provider: attributes_1.default.String({
        modelKey: 'provider',
    }),
    autocreate: attributes_1.default.Object({
        modelKey: 'autocreate',
    }),
};
