// nylas/lib/nylas.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
// TODO since node 10 URL is global
var url_1 = require("url");
var node_fetch_1 = __importDefault(require("node-fetch"));
var config = __importStar(require("./config"));
var nylas_connection_1 = __importDefault(require("./nylas-connection"));
var management_account_1 = __importDefault(require("./models/management-account"));
var account_1 = __importDefault(require("./models/account"));
var connect_1 = __importDefault(require("./models/connect"));
var restful_model_collection_1 = __importDefault(require("./models/restful-model-collection"));
var management_model_collection_1 = __importDefault(require("./models/management-model-collection"));
var webhook_1 = __importDefault(require("./models/webhook"));
var Nylas = /** @class */ (function () {
    function Nylas() {
    }
    Object.defineProperty(Nylas, "clientSecret", {
        get: function () {
            return config.clientSecret;
        },
        set: function (newClientSecret) {
            config.setClientSecret(newClientSecret);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Nylas, "apiServer", {
        get: function () {
            return config.apiServer;
        },
        set: function (newApiServer) {
            config.setApiServer(newApiServer);
        },
        enumerable: false,
        configurable: true
    });
    Nylas.config = function (_a) {
        var clientId = _a.clientId, clientSecret = _a.clientSecret, apiServer = _a.apiServer;
        if (apiServer && apiServer.indexOf('://') === -1) {
            throw new Error('Please specify a fully qualified URL for the API Server.');
        }
        if (clientId) {
            this.clientId = clientId;
        }
        if (clientSecret) {
            this.clientSecret = clientSecret;
        }
        if (apiServer) {
            this.apiServer = apiServer;
        }
        else {
            this.apiServer = 'https://api.nylas.com';
        }
        var conn = new nylas_connection_1.default(this.clientSecret, {
            clientId: this.clientId,
        });
        this.connect = new connect_1.default(conn, this.clientId, this.clientSecret);
        this.webhooks = new management_model_collection_1.default(webhook_1.default, conn, this.clientId);
        if (this.clientCredentials()) {
            this.accounts = new management_model_collection_1.default(management_account_1.default, conn, this.clientId);
        }
        else {
            this.accounts = new restful_model_collection_1.default(account_1.default, conn);
        }
        return this;
    };
    Nylas.clientCredentials = function () {
        return this.clientId != null && this.clientSecret != null;
    };
    Nylas.with = function (accessToken) {
        if (!accessToken) {
            throw new Error('This function requires an access token');
        }
        return new nylas_connection_1.default(accessToken, { clientId: this.clientId });
    };
    //TODO::Deprecate the camel case property in the next major release
    Nylas.application = function (options) {
        if (!this.clientId) {
            throw new Error('This function requires a clientId');
        }
        if (!this.clientSecret) {
            throw new Error('This function requires a clientSecret');
        }
        var connection = new nylas_connection_1.default(null, { clientId: this.clientId });
        var requestOptions = {
            path: "/a/" + this.clientId,
        };
        if (options) {
            requestOptions.body = {
                application_name: options.applicationName || options.application_name,
                redirect_uris: options.redirectUris || options.redirect_uris,
            };
            requestOptions.method = 'PUT';
        }
        return connection.request(requestOptions);
    };
    Nylas.exchangeCodeForToken = function (code, callback) {
        if (!this.clientId || !this.clientSecret) {
            throw new Error('exchangeCodeForToken() cannot be called until you provide a clientId and secret via config()');
        }
        if (!code) {
            throw new Error('exchangeCodeForToken() must be called with a code');
        }
        var url = new url_1.URL(this.apiServer + "/oauth/token");
        url.searchParams.set('client_id', this.clientId);
        url.searchParams.set('client_secret', this.clientSecret);
        url.searchParams.set('grant_type', 'authorization_code');
        url.searchParams.set('code', code);
        return node_fetch_1.default(url)
            .then(function (response) {
            if (response.ok) {
                return response.json();
            }
        })
            .then(function (body) {
            if (!body || !body['access_token']) {
                var errorMessage = 'No access token in response';
                if (body && body.message)
                    errorMessage = body.message;
                throw new Error(errorMessage);
            }
            if (callback) {
                callback(null, body['access_token']);
            }
            return body['access_token'];
        }, function (error) {
            var newError = new Error(error.message);
            if (callback) {
                callback(newError);
            }
            throw newError;
        });
    };
    Nylas.urlForAuthentication = function (options) {
        if (options === void 0) { options = {}; }
        if (!this.clientId) {
            throw new Error('urlForAuthentication() cannot be called until you provide a clientId via config()');
        }
        if (!options.redirectURI) {
            throw new Error('urlForAuthentication() requires options.redirectURI');
        }
        if (!options.loginHint) {
            options.loginHint = '';
        }
        var url = this.apiServer + "/oauth/authorize?client_id=" + this.clientId + "&response_type=code&login_hint=" + options.loginHint + "&redirect_uri=" + options.redirectURI;
        if (options.state != null) {
            url += "&state=" + options.state;
        }
        if (options.scopes != null) {
            url += "&scopes=" + options.scopes.join(',');
        }
        if (options.provider != null) {
            url += "&provider=" + options.provider;
        }
        return url;
    };
    Nylas.clientId = '';
    return Nylas;
}());
module.exports = Nylas;
