// nylas/lib/models/event-participant.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var restful_model_1 = __importDefault(require("./restful-model"));
var attributes_1 = __importDefault(require("./attributes"));
var EventParticipant = /** @class */ (function (_super) {
    __extends(EventParticipant, _super);
    function EventParticipant() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    EventParticipant.prototype.toJSON = function (enforceReadOnly) {
        var json = _super.prototype.toJSON.call(this, enforceReadOnly);
        if (!json['name']) {
            json['name'] = json['email'];
        }
        return json;
    };
    return EventParticipant;
}(restful_model_1.default));
exports.default = EventParticipant;
EventParticipant.collectionName = 'event-participants';
EventParticipant.attributes = {
    name: attributes_1.default.String({
        modelKey: 'name',
    }),
    email: attributes_1.default.String({
        modelKey: 'email',
    }),
    status: attributes_1.default.String({
        modelKey: 'status',
        readOnly: true,
    }),
};
